### ISSUE TITLE
Write bootloader - problem on ATmega8

### ISSUE THREAD

**Issue by Testato** [arduino/Arduino#2075]
During the writing of the bootloader, IDE 1.5.x look for e_fuse setting for the ATmega8 (ArduinoNG board) 
But the ATmega8 do not have the e_fuse, so it is normal that there isn't

[code]Arduino: 1.5.4 (Windows 7), Board: "Arduino NG or older, ATmega8"
"efuse" memory type not defined for part "ATMEGA8"[/code]

I also tried to manually add a fake e_fuse in the board, without success 
[code]atmegang.menu.cpu.atmega8.bootloader.extended_fuses=0xFF
[/code]

**Comment by rigelinorion** [comment]
Hello, I solved it this way...

Open for edit the file C:\Program Files\Arduino\hardware\arduino\avr\platforms.txt
This is a standard section programmers tools...
==============================================================================
# AVR Uploader/Programmers tools
# ------------------------------
tools.avrdude.cmd.path={runtime.ide.path}/hardware/tools/avr/bin/avrdude
tools.avrdude.config.path={runtime.ide.path}/hardware/tools/avr/etc/avrdude.conf
tools.avrdude.cmd.path.linux={runtime.ide.path}/hardware/tools/avrdude
tools.avrdude.config.path.linux={runtime.ide.path}/hardware/tools/avrdude.conf

tools.avrdude.upload.params.verbose=-v -v -v -v
tools.avrdude.upload.params.quiet=-q -q
tools.avrdude.upload.pattern="{cmd.path}" "-C{config.path}" {upload.verbose} -p{build.mcu} -c{upload.protocol} -P{serial.port} -b{upload.speed} -D "-Uflash:w:{build.path}/{build.project_name}.hex:i"

tools.avrdude.program.params.verbose=-v -v -v -v
tools.avrdude.program.params.quiet=-q -q
tools.avrdude.program.pattern="{cmd.path}" "-C{config.path}" {program.verbose} -p{build.mcu} -c{protocol} {program.extra_params} "-Uflash:w:{build.path}/{build.project_name}.hex:i"

tools.avrdude.erase.params.verbose=-v -v -v -v
tools.avrdude.erase.params.quiet=-q -q
tools.avrdude.erase.pattern="{cmd.path}" "-C{config.path}" {erase.verbose} -p{build.mcu} -c{protocol} {program.extra_params} -e -Uefuse:w:{bootloader.extended_fuses}:m -Ulock:w:{bootloader.unlock_bits}:m  -Uhfuse:w:{bootloader.high_fuses}:m -Ulfuse:w:{bootloader.low_fuses}:m


tools.avrdude.bootloader.params.verbose=-v -v -v -v
tools.avrdude.bootloader.params.quiet=-q -q
tools.avrdude.bootloader.pattern="{cmd.path}" "-C{config.path}" {bootloader.verbose} -p{build.mcu} -c{protocol} {program.extra_params} "-Uflash:w:{runtime.ide.path}/hardware/arduino/avr/bootloaders/{bootloader.file}:i" -Ulock:w:{bootloader.lock_bits}:m
# ------------------------------
==============================================================================

create a new tool called, for example, avrdude1
simply copy section tools.avrdude and name it tools.avrdude1
the trick is that we remove switch "-Uefuse:w:{bootloader.extended_fuses}:m" from  "tools.avrdude1.erase.pattern="

==============================================================================
# ------------------------------
tools.avrdude1.cmd.path={runtime.ide.path}/hardware/tools/avr/bin/avrdude
tools.avrdude1.config.path={runtime.ide.path}/hardware/tools/avr/etc/avrdude.conf
tools.avrdude1.cmd.path.linux={runtime.ide.path}/hardware/tools/avrdude
tools.avrdude1.config.path.linux={runtime.ide.path}/hardware/tools/avrdude.conf

tools.avrdude1.upload.params.verbose=-v -v -v -v
tools.avrdude1.upload.params.quiet=-q -q
tools.avrdude1.upload.pattern="{cmd.path}" "-C{config.path}" {upload.verbose} -p{build.mcu} -c{upload.protocol} -P{serial.port} -b{upload.speed} -D "-Uflash:w:{build.path}/{build.project_name}.hex:i"

tools.avrdude1.program.params.verbose=-v -v -v -v
tools.avrdude1.program.params.quiet=-q -q
tools.avrdude1.program.pattern="{cmd.path}" "-C{config.path}" {program.verbose} -p{build.mcu} -c{protocol} {program.extra_params} "-Uflash:w:{build.path}/{build.project_name}.hex:i"

tools.avrdude1.erase.params.verbose=-v -v -v -v
tools.avrdude1.erase.params.quiet=-q -q

#this line was edited
tools.avrdude1.erase.pattern="{cmd.path}" "-C{config.path}" {erase.verbose} -p{build.mcu} -c{protocol} {program.extra_params} -e -Ulock:w:{bootloader.unlock_bits}:m  -Uhfuse:w:{bootloader.high_fuses}:m -Ulfuse:w:{bootloader.low_fuses}:m

tools.avrdude1.bootloader.params.verbose=-v -v -v -v
tools.avrdude1.bootloader.params.quiet=-q -q
tools.avrdude1.bootloader.pattern="{cmd.path}" "-C{config.path}" {bootloader.verbose} -p{build.mcu} -c{protocol} {program.extra_params} "-Uflash:w:{runtime.ide.path}/hardware/arduino/avr/bootloaders/{bootloader.file}:i" -Ulock:w:{bootloader.lock_bits}:m
# ------------------------------
==============================================================================

Here is an example of the use of the tool avrdude1 in boards.txt file

==============================================================================

atmega8.name=ATmega8 OPTIBOOT
atmega8.upload.protocol=arduino
atmega8.upload.maximum_size=7168
atmega8.upload.speed=115200
atmega8.upload.tool=avrdude1
atmega8.bootloader.low_fuses=0xbf
atmega8.bootloader.high_fuses=0xCC
atmega8.bootloader.path=atmega8
atmega8.bootloader.file=optiboot/optiboot_atmega8.hex
atmega8.bootloader.unlock_bits=0x3F
atmega8.bootloader.lock_bits=0x0F
atmega8.bootloader.tool=avrdude1
atmega8.build.mcu=atmega8
atmega8.build.f_cpu=16000000L
atmega8.build.core=arduino
atmega8.build.variant=standard

**Comment by testato** [comment]
Thanks for this tips.
But this cannot be the final version because the customer should not create
new board or change platform file. He must only choose ArduinoNG from
Arduino IDE and burn bootloader on Atmega8. Because Atmega8 is official Mcu
on an official Arduino board
Il 08/gen/2015 09:06 "rigelinorion" <notifications@github.com> ha scritto:

**Comment by rigelinorion** [comment]
So you can create a batch file that edits files and platforms.txt boards.txt. Or maybe the point is that in general you can not change the contents of these files?

**Comment by testato** [comment]
A people that have an Original ArduinoNG board must receive an IDE that
work out of the box.
After this point he may, or not, try to understand how arduino ide work and
modify/learn from the code.
So your solution is very important, and now at least is there a manual
solution on this bug, but the Team must not close this Issue, this issue
will closed when, trasparently, the Arduino IDE 1.5.x will work on
ArduinoNG board
Il 08/gen/2015 09:29 "rigelinorion" <notifications@github.com> ha scritto:

**Comment by rigelinorion** [comment]
I understand. Thank you for your response testato.

**Comment by DRSDavidSoft** [comment]
When this bug will be fixed?

**Comment by joseluisperezbarrales** [comment]
Super rigelinorion !!!
Great, great great, Txs a lot. It was so simple...
Cojonudo. Va perfecto

**Comment by sudiptamandal** [comment]
how to create a new tool called, avrdude1?? Help me creating new tool

**Comment by DRSDavidSoft** [comment]
You should edit the file platform.txt file, usually found in:
arduino\hardware\arduino\avr\platform.txt

Also, on windows you should probably use Notepad++, Metapad or similar programs to edit the files.

BTW, I suggest to name it avrdude_old or avrdude_noefuse other than avrdude1.

**Comment by Testato** [comment]
One year for a Regression bug is not acceptable.

**Comment by GiovanniCmp** [comment]
Any progress in this?

**Comment by RomanoLindano** [comment]
I wait for this bug to be resolved.

**Comment by Testato** [comment]
:+1:

**Comment by PaoloP74** [comment]
+1

**Comment by GrimKriegor** [comment]
+1 using v1.6.5

**Comment by Ashum812** [comment]
Sure do the solution by "rigelinorion"  works bootloader uploaded succeccfully  but i'm still not able to upload sketch to the atmega8 it says

avrdude: stk500_recv(): programmer is not responding
avrdude: stk500_getsync() attempt 10 of 10: not in sync: resp=0xd3

**Comment by GrimKriegor** [comment]
@Ashum812 have you tried "Upload using the programmer"?

**Comment by rigelinorion** [comment]
Perhaps the fault of the bootloader. I use optiboot and everything works well. Maybe fusbits are not set properly? Is reset is triggered when uploading? What programmer you are trying to upload the program?

**Comment by Testato** [comment]
Pay attention that "Upload using the programmer" remove the bootloader, so it's use is completely out of topic respect this issue, that ask (from one years and an half), to resolve the bug on the ArduinoNG official board.

**Comment by wasimxe** [comment]
I am using Arduino 1.6.8 but still this error :(

**Comment by wasimxe** [comment]
Modifying rigelinorion comment, Do whatever rigelinorion says and then 

atmega8_diy_atmega8a_8MHz_optiboot.name=ATmega8A, (8MHz Internal Oscillator)
atmega8_diy_atmega8a_8MHz_optiboot.build.core=arduino:arduino
atmega8_diy_atmega8a_8MHz_optiboot.build.variant=arduino:standard
atmega8_diy_atmega8a_8MHz_optiboot.build.board=AVR_NG
atmega8_diy_atmega8a_8MHz_optiboot.build.mcu= atmega8
atmega8_diy_atmega8a_8MHz_optiboot.build.f_cpu=8000000L
atmega8_diy_atmega8a_8MHz_optiboot.bootloader.tool=arduino:avrdude1
atmega8_diy_atmega8a_8MHz_optiboot.bootloader.unlock_bits=0x3F
atmega8_diy_atmega8a_8MHz_optiboot.bootloader.lock_bits=0x0F
atmega8_diy_atmega8a_8MHz_optiboot.bootloader.low_fuses=0xA4
atmega8_diy_atmega8a_8MHz_optiboot.bootloader.high_fuses=0xCC
atmega8_diy_atmega8a_8MHz_optiboot.bootloader.file=atmega8/optiboot_atmega8_8000000L_57600.hex
atmega8_diy_atmega8a_8MHz_optiboot.upload.maximum_size=7168
atmega8_diy_atmega8a_8MHz_optiboot.upload.maximum_data_size=1024
atmega8_diy_atmega8a_8MHz_optiboot.upload.tool=avrdude
atmega8_diy_atmega8a_8MHz_optiboot.upload.protocol=arduino
atmega8_diy_atmega8a_8MHz_optiboot.upload.speed=19200

Actually problem with rigelinorion was it can bootload but no upload sketch, so for bootloader use tool avrdude1 and for uploading using build-in Atmega8 NG, above one is combination of both in one.

**Comment by Testato** [comment]
Happy birthday !!!
Still open after two years from the opening of the issue.
This is a historical record.
@cmaglie @mbanzi

**Comment by DRSDavidSoft** [comment]
@cmaglie **Maximum Effort**

**Comment by brainstorm** [comment]
I think that the problem might be upstream with avrdude's default configuration, since it does not seem to define atmega8's boot/efuse memory regions in `avrdude.conf`:

```
#------------------------------------------------------------
# ATmega8
#------------------------------------------------------------

part
    id               = "m8";
    desc             = "ATmega8";
    stk500_devcode   = 0x70;
    avr910_devcode   = 0x76;
    signature        = 0x1e 0x93 0x07;
    pagel            = 0xd7;
    bs2              = 0xc2;
    chip_erase_delay = 10000;
    pgm_enable       = "1 0 1 0  1 1 0 0    0 1 0 1  0 0 1 1",
                       "x x x x  x x x x    x x x x  x x x x";

    chip_erase       = "1 0 1 0  1 1 0 0    1 0 0 x  x x x x",
                       "x x x x  x x x x    x x x x  x x x x";

    timeout             = 200;
    stabdelay           = 100;
    cmdexedelay         = 25;
    synchloops          = 32;
    bytedelay           = 0;
    pollindex           = 3;
    pollvalue           = 0x53;
    predelay            = 1;
    postdelay           = 1;
    pollmethod          = 0;

    pp_controlstack     =
        0x0E, 0x1E, 0x0F, 0x1F, 0x2E, 0x3E, 0x2F, 0x3F,
        0x4E, 0x5E, 0x4F, 0x5F, 0x6E, 0x7E, 0x6F, 0x7F,
        0x66, 0x76, 0x67, 0x77, 0x6A, 0x7A, 0x6B, 0x7B,
        0xBE, 0xFD, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00;
    hventerstabdelay    = 100;
    progmodedelay       = 0;
    latchcycles         = 5;
    togglevtg           = 1;
    poweroffdelay       = 15;
    resetdelayms        = 2;
    resetdelayus        = 0;
    hvleavestabdelay    = 15;
    resetdelay          = 15;
    chiperasepulsewidth = 0;
    chiperasepolltimeout = 10;
    programfusepulsewidth = 0;
    programfusepolltimeout = 5;
    programlockpulsewidth = 0;
    programlockpolltimeout = 5;

    memory "eeprom"
        size            = 512;
        page_size       = 4;
        min_write_delay = 9000;
(...)
    memory "flash"
        paged           = yes;
        size            = 8192;
        page_size       = 64;
        num_pages       = 128;
        min_write_delay = 4500;
        max_write_delay = 4500;
        readback_p1     = 0xff;
        readback_p2     = 0x00;
(...)
    memory "lfuse"
        size            = 1;
        min_write_delay = 2000;
        max_write_delay = 2000;
        read            = "0 1 0 1  0 0 0 0   0 0 0 0  0 0 0 0",
                          "x x x x  x x x x   o o o o  o o o o";
    memory "hfuse"
        size            = 1;
        min_write_delay = 2000;
        max_write_delay = 2000;
        read            = "0 1 0 1  1 0 0 0   0 0 0 0  1 0 0 0",
                          "x x x x  x x x x   o o o o  o o o o";

        write           = "1 0 1 0  1 1 0 0   1 0 1 0  1 0 0 0",
                          "x x x x  x x x x   i i i i  i i i i";
      ;

    memory "lock"
        size            = 1;
        min_write_delay = 2000;
        max_write_delay = 2000;
        read            = "0 1 0 1  1 0 0 0   0 0 0 0  0 0 0 0",
                          "x x x x  x x x x   x x o o  o o o o";

        write           = "1 0 1 0  1 1 0 0   1 1 1 x  x x x x",
                          "x x x x  x x x x   1 1 i i  i i i i";
      ;

    memory "calibration"
        size            = 4;
        read            = "0  0  1  1   1  0  0  0   0  0  x  x   x  x  x  x",
                          "0  0  0  0   0  0 a1 a0   o  o  o  o   o  o  o  o";
      ;

    memory "signature"
        size            = 3;
        read            = "0  0  1  1   0  0  0  0   x  x  x  x   x  x  x  x",
                          "x  x  x  x   x  x a1 a0   o  o  o  o   o  o  o  o";
      ;
```

Therefore, i.e:

```
root@kali:~/ardrone_motors/motor2# avrdude -p m8 -C /etc/avrdude.conf -c ardrone_motor_2 -U boot:r:boot.hex:i

avrdude: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude: Device signature = 0x1e9307 (probably m8)
"boot" memory type not defined for part "ATmega8"

avrdude: safemode: Fuses OK (E:FF, H:DC, L:E4)

avrdude done.  Thank you.
```

**Comment by RohithPallerla** [comment]
My suggestion which worked out for me was to use other versions of arduino.
In my case, I have used an older version- Arduino alpha
Also working with newer versions.


Hope this would be helpful:)

**Comment by Testato** [comment]
There is many solution, from modify the IDE to use old version of the ide.
But the problem is that the arduino team do not have time, or do not want, solve this problem.
This is unacceptable !

**Comment by awatterott** [comment]
Fix: add fake efuse to ATmega8 in avrdude.conf. Tested and works with avrdude 6.0.1 and 6.3.
 ```
    memory "efuse"
        size            = 0;
      ;
```
https://github.com/watterott/wattuino/commit/140b04486691d1eebeb4f8f84e4e3fa5b593c970

**Comment by Testato** [comment]
avrdude.conf is a modified arduino file ? if yes you can prepare a PR end solve this storic bug.
But if avrdude.conf is a file that the arduino team do not manage, because it want use the original file from avrdude team, this is not the solution

**Comment by awatterott** [comment]
The file comes from the avrdude project and can be patched [during building](https://github.com/arduino/avrdude-build-script) from the Arduino team.

**Comment by Testato** [comment]
> As soon as avrdude ships a new version, we pull the source code, patch it with some user contributed patches and deliver it with the Arduino IDE. 

So it is possible solve this bug, but how "contribute a patches" that arduino team will store and copy at every new avrdude update ?

If you know how please create this patch, we organize a party if after "only" two years this bug will solved.

**Comment by aakrutii** [comment]
i have solved this problem check this link    @Testato @diegocueva @rigelinorion @DRSDavidSoft @joseluisperezbarrales 

https://www.youtube.com/watch?v=2MMYnIFaXv4

you will get the solution here

**Comment by awatterott** [comment]
@aakrutii This is not a solution. You have removed the efuse parameter, so that the bootloader burning of all microcontrollers with an extended fuse will not work.
I have posted a workaround [here](https://github.com/arduino/Arduino/issues/2075#issuecomment-238031689).

**Comment by DRSDavidSoft** [comment]
If what @awatterott says is true, this would break the way Arduinos with Atmega328p can be bootloaded. I like his workaround better.

**Comment by cmaglie** [comment]
I've pushed a fix as suggested by @awatterott:

https://github.com/arduino/avrdude-build-script/commit/0bfe2c433ffe9b0371c5ac1adf5bf9ecf812cf12
https://github.com/arduino/Arduino/commit/b1f349244b85bd8c39c753dc0e267a1e7a5f01c6

It should be already in the hourly build, if anyone wants to try it.

:santa:

**Comment by cmaglie** [comment]
Should be fixed from 1.6.14. Please reopen if the issue persists.

**Comment by Testato** [comment]
that's for sure :-)
